### PR TITLE
[None][fix] fix Cosmos3 multi-control attention semantics

### PR DESCRIPTION
**Pull request by ishovkun**
## Summary

- run Cosmos3 attention independently for every `[control_i, target]` pair
- preserve each control output while combining target outputs with normalized relative weights
- preserve the existing single-control fast path
- keep multi-control generator layers sequence-sharded under pure Ulysses, including TP + Ulysses

## Computation

For each generator layer and control `i`:

```text
result_i = Attention(
    Q=[control_i, target],
    K=[text, control_i, target],
    V=[text, control_i, target],
)

next_control_i = result_i.control_output
next_target = sum_i(normalized_weight_i * result_i.target_output)
```

Weights default to uniform mixing and are validated for matching length, finiteness, non-negativity, and a positive total. Explicit internal weights are normalized once before layer execution. `control_guidance` remains the existing outer control-CFG operation and is not used as a per-control weight.

## Ulysses handling

For pure Ulysses, the full generator is no longer replicated per rank:

1. `hidden_gen`, RoPE tensors, and cached text K/V remain sequence-sharded through the generator layers.
2. Each layer performs the normal Ulysses sequence-to-head exchange for Q/K/V, restoring the global sequence while retaining sharded attention heads.
3. Rank-packed K/V is sliced into independent `[text, control_i, target]` inputs. The inner configured backend runs once per control pair.
4. Control outputs are preserved, target outputs are weight-normalized and summed, and the reconstructed generator sequence is exchanged back to sequence shards.

This restores Ulysses head and sequence parallelism while preserving the required independent attention equation. The `N` independent attention calls are inherent to the official multi-control computation; this change removes the additional full-generator replication and separately pinned-backend penalties. TP2 + Ulysses2 parity is covered against a non-parallel reference. Combined Attention2D/context-parallel compositions retain the guarded replicated fallback; the intended transfer layout (`cfg_size=2`, `ulysses_size=4`) is pure Ulysses.

## Behavioral reference

The implementation follows NVIDIA's Cosmos-framework multi-control path:

- [`CrossAttention._forward_multi_control`](https://github.com/NVIDIA/cosmos-framework/blob/d61cbe96e8a9b6ce22e92268f29dcd5876f6c417/cosmos_framework/model/generator/mot/attention.py#L428-L579) runs independent control/target attention and combines only target outputs by weight.
- [Cosmos3 VFM Ulysses handling](https://github.com/NVIDIA/cosmos-framework/blob/d61cbe96e8a9b6ce22e92268f29dcd5876f6c417/cosmos_framework/model/generator/mot/cosmos3_vfm_network.py#L1178-L1215) restores the global sequence with Ulysses all-to-all while keeping attention heads sharded.
- [Transfer weight preparation](https://github.com/NVIDIA/cosmos-framework/blob/d61cbe96e8a9b6ce22e92268f29dcd5876f6c417/cosmos_framework/inference/transfer.py#L132-L176) supplies normalized per-control weights.
- NVIDIA's official [edge + blur recipe](https://github.com/NVIDIA/cosmos/blob/337c1f44872c243db1e3f396037597b946ea6268/cookbooks/cosmos3/generator/transfer/specs/multi_control.json#L22-L28) assigns `0.5` to each control.

## Testing

- `7 passed`: focused multi-control equation, isolation, weight, smoke, and Ulysses range tests
- `2 passed`: multi-control Ulysses2 and TP2 + Ulysses2 parity against a non-parallel reference on 4x H200
- `1 passed`: ordinary Ulysses2 regression against a non-parallel reference
- broader Cosmos3 transfer/transformer suite: `119 passed`; one unrelated checkpoint FP8 smoke reproducibly emitted all-NaN output without entering multi-control or Ulysses code
- applicable pre-commit checks passed

No benchmark client, serving wrapper, public API, or runbook changes are included. An end-to-end edge + blur MP4 run was not performed in this update because the adopted allocation has 4 GPUs rather than the intended 8-GPU layout and no caller-supplied control videos were provided.

## Dev Engineer Review

- Added Cosmos3 multi-control attention for independent `[control_i, target]` pairs.
- Preserved control outputs and combined target outputs with validated normalized weights.
- Preserved single-control fast-path behavior.
- Added lazy construction of the vanilla fallback backend.
- Added reusable Ulysses sequence/head conversion helpers.
- Updated forward signatures for multi-control metadata.
- Preserved sequence sharding for pure Ulysses and TP + Ulysses.
- Retained the guarded replicated fallback for combined Attention2D/context-parallel configurations.
- No configuration, test-list, serving, benchmark, runbook, or external public API changes were identified.

## QA Engineer Review

Added coverage in:

- `tests/unittest/_torch/visual_gen/test_cosmos3_transformer.py`
  - Multi-control forward behavior and control-token isolation.
  - Uniform and weighted aggregation.
  - Invalid control-weight validation.
  - Multi-head Q/K/V assertions.
  - Per-item text-prefix lengths.
  - Single-control path selection.
- `tests/unittest/_torch/visual_gen/multi_gpu/test_cosmos3_transformer_parallel.py`
  - `test_multi_control_ulysses2_vs_single_gpu`
  - `test_multi_control_tp2_ulysses2_vs_single_gpu`

The unit test file is listed in `tests/integration/test_lists/test-db/l0_b200.yml`. The multi-GPU test file is listed in `tests/integration/test_lists/test-db/l0_dgx_b200.yml`.

Validation passed for focused multi-control tests, Ulysses2 and TP2 + Ulysses2 parity tests, broader Cosmos3 coverage, and pre-commit checks.

**Verdict: sufficient.**

### CODE REVIEW

**Comment by ishovkun** [comment]
@coderabbitai Addressed both nitpicks in `73a7658042`:

- Removed the unreachable post-loop coverage check. Weighted target outputs are now collected and reduced in control order without the dead `RuntimeError` branch.
- Expanded the mocked multi-control coverage to two attention heads with `head_dim=2`, with exact Q/K/V assertions that distinguish sequence, head, and feature axes. Added a two-item batch with `real_text_lens=[1, 2]` to exercise the per-item text-prefix path.

Validation:

- `10 passed, 19 deselected` for `TestCosmos3MultiControlAttention` on B300
- pre-commit passed for both modified files

Please re-review the update.

File: tensorrt_llm/_torch/visual_gen/models/cosmos3/transformer_cosmos3.py [hunk 1]
```diff
@@ -542,11 +570,129 @@ def __init__(
...
+        out = self.multi_control_attn.forward(
+            q=q.transpose(1, 2),
+            k=k.transpose(1, 2),
+            v=v.transpose(1, 2),
+            attention_mask=PredefinedAttentionMask.FULL,
+            timestep=timestep,
+        )
+        return out.transpose(1, 2)
+
+    def _forward_multi_control(
```
Comment (karljang) [inline tensorrt_llm/_torch/visual_gen/models/cosmos3/transformer_cosmos3.py:734]: **New multi-control semantics have no cited reference, and the E2E guards are waived**

What defines "correct" here? The tests assert isolation, weighting, finiteness, and parallel-vs-single-GPU parity — but both sides of each comparison run this new formulation, so they would pass equally if the formula itself were wrong. Could you cite the upstream/reference implementation that `result_i = Attention(Q=[control_i, target], K=V=[text, control_i, target])` matches? That is what turns this from a behaviour change into a fix.

This is load-bearing right now because the usual safety net is off: all three Cosmos3 accuracy guards are waived on main — `test_cosmos3_feature_accuracy_against_golden[nvfp4]` (nvbugs/6572800), `test_cosmos3_nano_t2v_lpips_against_golden` and `test_cosmos3_nano_v2v_lpips_against_golden` (both nvbugs/6655359) — and `control_latents`/`transfer` appear nowhere in `test_visual_gen_cosmos3.py`, so there is no multi-control E2E coverage at all.

File: tensorrt_llm/_torch/visual_gen/models/cosmos3/transformer_cosmos3.py [hunk 2]
```diff
@@ -1410,15 +1592,19 @@ def forward(
...
 
         S_gen = hidden_gen.shape[1]
-        hidden_gen = self.sharder.shard(hidden_gen, dim=1, pad_to_multiple=True)
+        control_token_sizes = tuple(control.shape[1] for control in hidden_controls)
+        multi_control_token_sizes = control_token_sizes if use_multi_control_attention else None
+        multi_control_weights = normalized_control_weights if use_multi_control_attention else None
         cos, sin = freqs_gen_combined
-        cos = self.sharder.shard(cos, dim=1, pad_to_multiple=True)
-        sin = self.sharder.shard(sin, dim=1, pad_to_multiple=True)
+        if not use_multi_control_attention:
```
Comment (karljang) [inline tensorrt_llm/_torch/visual_gen/models/cosmos3/transformer_cosmos3.py]: **Multi-control pays three compounding costs, documented only by warning_once**

Multi-control now (a) skips the `hidden_gen`/`cos`/`sin` sharding and the final `gather`, (b) runs N separate attentions instead of one, and (c) is pinned to the `VANILLA` backend. For N controls of size C and target length T with T dominant, the attention cost goes from roughly `(NC+T)(text+NC+T)` to `N(C+T)(text+C+T)` — about **N× the FLOPs** — on top of losing sequence parallelism and the optimized backend.

That may well be the right correctness-first trade, but a `logger.warning_once` is the only place it is recorded. Could you add a measured cost for a representative N, or state explicitly in the PR description that multi-control is correctness-first for now?

File: tensorrt_llm/_torch/visual_gen/models/cosmos3/transformer_cosmos3.py [hunk 3]
```diff
@@ -542,11 +570,129 @@ def __init__(
         eps = model_config.pretrained_config.rms_norm_eps
         self.norm_q = NemotronRMSNorm(hidden_size=head_dim, eps=eps, dtype=torch.bfloat16)
         self.norm_k = NemotronRMSNorm(hidden_size=head_dim, eps=eps, dtype=torch.bfloat16)
+        # Multi-control pairs stay replicated across sequence-parallel ranks;
+        # this backend deliberately has no Ulysses/Ring/Attention2D wrapper.
+        self.multi_control_attn = create_attention(
```
Comment (karljang) [inline tensorrt_llm/_torch/visual_gen/models/cosmos3/transformer_cosmos3.py]: **multi_control_attn is constructed unconditionally in every layer**

`self.multi_control_attn` is built in each layer's `__init__` regardless of whether multi-control is ever used, so single-control and no-control models allocate one unused attention module per layer. Cheap, but constructing it lazily on first multi-control use would avoid it entirely.

File: tensorrt_llm/_torch/visual_gen/models/cosmos3/transformer_cosmos3.py [hunk 4]
```diff
@@ -28,6 +28,7 @@
 from tensorrt_llm._torch.modules.linear import Linear, WeightMode
 from tensorrt_llm._torch.modules.mlp import MLP
 from tensorrt_llm._torch.utils import relu2
+from tensorrt_llm._torch.visual_gen.attention_backend.utils import create_attention
```
Comment (karljang) [inline tensorrt_llm/_torch/visual_gen/models/cosmos3/transformer_cosmos3.py:37]: **Two sibling PRs touch this file and both test files — agree on merge order**

Worth coordinating merge order: #17449 (visual_gen attention-handling refactor) touches `attention_backend/utils.py` — the `create_attention` newly imported here — plus this file and **both** test files this PR edits. #17325 (your Cosmos3 action generation) also touches this file and both test files. #17449 is the one that could invalidate the new `create_attention(backend="VANILLA", ...)` call site.

File: tests/unittest/_torch/visual_gen/test_cosmos3_transformer.py [hunk 5]
```diff
@@ -229,6 +231,44 @@ def test_sanity_forward(self, cosmos3_model_config):
             )
         _assert_finite_output(out.video, hs.shape)
 
+    @pytest.mark.high_cuda_memory
+    def test_sanity_forward_multi_control(self, cosmos3_model_config):
```
Comment (karljang) [inline tests/unittest/_torch/visual_gen/test_cosmos3_transformer.py]: **CodeRabbit's QA verdict is a false positive — these tests already run**

Ignore the automated QA verdict asking for test-list coverage. Both files already exist on main and are registered by path — `l0_b200.yml:265` for this file and `l0_dgx_b200.yml:209` for the multi-GPU one — and registration is per file, not per test function, so the new tests are collected automatically. No test-list change is needed.

**Review by karljang — COMMENTED** [review]
Reviewed the diff and verified what I could against the tree at `73a7658`. The implementation is coherent and the weight validation is thorough. Two things I would like to resolve before this lands, plus one automated finding you can safely ignore — inline comments below.

The main question is what defines "correct" here. The tests establish that the new path is internally consistent and parallel-safe, but both sides of every comparison run the new semantics, so nothing here shows the new formulation is the right one. That matters more than usual right now, because all three Cosmos3 accuracy guards are currently waived on main.

The performance note is a heads-up on cost rather than an objection.

Confirmed as correct, so no action needed: preserving per-pair control outputs does **not** conflict with the "their outputs are discarded" docstring, since `transformer_cosmos3.py:1447` strips the control prefix before the final projection.

What I did not verify: I did not run the tests, so the "107 passed / 2 passed on B300" numbers and the parity tolerances are unchecked.

File: tensorrt_llm/_torch/visual_gen/models/cosmos3/transformer_cosmos3.py (hunk 1 above)
Comment (ishovkun) [reply tensorrt_llm/_torch/visual_gen/models/cosmos3/transformer_cosmos3.py:734]: Good point—the TRT-LLM tests establish operator behavior and parallel consistency, but they are not an independent semantic oracle.
This implementation ports vLLM-Omni’s merged Cosmos3 transfer fix, commit [b3270dec8d75](https://github.com/vllm-project/vllm-omni/commit/b3270dec8d758b9801aa0d01bd7205757b65b814). Its [_forward_multi_control](https://github.com/vllm-project/vllm-omni/blob/b3270dec8d758b9801aa0d01bd7205757b65b814/vllm_omni/diffusion/models/cosmos3/transformer_cosmos3.py#L717-L770) runs one attention pass per [control_i, target], preserves each corresponding control output, and combines only the target outputs using normalized weights. The same commit defines [uniform default weights and normalization](https://github.com/vllm-project/vllm-omni/blob/b3270dec8d758b9801aa0d01bd7205757b65b814/vllm_omni/diffusion/models/cosmos3/transformer_cosmos3.py#L1725-L1748).
NVIDIA’s official [Cosmos3 edge+blur recipe](https://github.com/NVIDIA/cosmos/blob/337c1f44872c243db1e3f396037597b946ea6268/cookbooks/cosmos3/generator/transfer/specs/multi_control.json#L22-L28) assigns weight 0.5 to each control.
I added these immutable references to the PR description. You are also correct about the current E2E gap: these tests cover the equation, isolation, tensor layout, and parallel parity, but they do not provide an independent multi-control quality golden.

File: tensorrt_llm/_torch/visual_gen/models/cosmos3/transformer_cosmos3.py (hunk 2 above)
Comment (ishovkun) [reply tensorrt_llm/_torch/visual_gen/models/cosmos3/transformer_cosmos3.py]: Good catch. The `N` independent attention calls are required by the [official Cosmos-framework multi-control computation](https://github.com/NVIDIA/cosmos-framework/blob/d61cbe96e8a9b6ce22e92268f29dcd5876f6c417/cosmos_framework/model/generator/mot/attention.py#L428-L579), but losing Ulysses sharding and pinning a separate vanilla backend were not.

Fixed in [d60dd6c47e](https://github.com/NVIDIA/TensorRT-LLM/pull/18277/commits/d60dd6c47e266cbf1d84ff3004650e67826e7695). Multi-control now keeps the generator sequence sharded, uses the normal Ulysses sequence-to-head exchange, runs each control pair through the configured inner backend with sharded heads, and exchanges the reconstructed output back to sequence shards.

Ulysses2 and TP2+Ulysses2 parity pass against the single-GPU reference, with tests that reject the old replicated fallback. I also updated the PR description to distinguish the inherent `N` attention cost from the removed replication/backend penalties.

File: tensorrt_llm/_torch/visual_gen/models/cosmos3/transformer_cosmos3.py (hunk 3 above)
Comment (ishovkun) [reply tensorrt_llm/_torch/visual_gen/models/cosmos3/transformer_cosmos3.py]: Agreed. Fixed in [b111b72225](https://github.com/NVIDIA/TensorRT-LLM/pull/18277/commits/b111b7222589afd4473028f0c845e40c6d0d1c25). `multi_control_attn` now starts as None and is constructed and cached only when the single-worker/context-parallel fallback is first used. No-control, single-control, and pure-Ulysses execution never construct it.
I added assertions covering both sides: the fallback is absent after ordinary model construction and is created on the first multi-control fallback execution.

File: tensorrt_llm/_torch/visual_gen/models/cosmos3/transformer_cosmos3.py (hunk 4 above)
Comment (ishovkun) [reply tensorrt_llm/_torch/visual_gen/models/cosmos3/transformer_cosmos3.py:37]: Thanks for flagging this. #[17325](https://github.com/NVIDIA/TensorRT-LLM/pull/17325) has already merged, and this PR is rebased onto it, so that overlap is resolved.
#[17449](https://github.com/NVIDIA/TensorRT-LLM/pull/17449) is still open. I checked its current diff: the create_attention(...) arguments used here remain supported, but #17449 changes the attention-metadata contract and the Cosmos3 call sites, so whichever PR lands second will need a deliberate rebase and multi-control metadata reconciliation.
I’ll sync with Ruqing on that interaction. If this PR lands first, I’ll help carry the multi-control path through #17449’s refactor; if #17449 lands first, I’ll rebase this PR and rerun the single-GPU, Ulysses, and TP+Ulysses coverage before merge.

File: tests/unittest/_torch/visual_gen/test_cosmos3_transformer.py (hunk 5 above)
Comment (ishovkun) [reply tests/unittest/_torch/visual_gen/test_cosmos3_transformer.py]: Thanks for checking. Confirmed both test files are already registered, and the test lists collect them at file scope, so the new test functions will run automatically. I’ll leave the test lists unchanged and disregard the automated QA verdict.

File: tests/unittest/_torch/visual_gen/test_cosmos3_transformer.py [hunk 6]
```diff
@@ -477,6 +517,189 @@ def test_load_fp8_quantization(self, quant_algo: str):
...
 
 
+class _RecordingMultiControlBackend:
+    def __init__(self, control_ids=(1, 2)):
+        self.calls = []
+        self.control_ids = iter(control_ids)
+
+    def forward(self, *, q, k, v, **_kwargs):
+        self.calls.append({"q": q.clone(), "k": k.clone(), "v": v.clone()})
+        control_id = next(self.control_ids)
```
[the remark this answers is not in the document]
Comment (ishovkun) [reply tests/unittest/_torch/visual_gen/test_cosmos3_transformer.py:824]: Addressed in [799d7d42ba](https://github.com/NVIDIA/TensorRT-LLM/pull/18277/commits/799d7d42ba). Added precise parameter and return annotations to the complete new multi-control test surface, including the unit-test backends/helpers, nested attention stub, test methods, and the distributed Ulysses helpers/tests. `py_compile` and all applicable pre-commit checks pass.
